editor/plugins/gradient_editor_plugin.cpp

```diff
@@ -31,27 +31,648 @@
+		undo_redo->add_undo_method(this, "_gradient_update");
+		undo_redo->commit_action();
+		editing = false;
+
+		// Reselect the old selected point if it's not the deleted one.
+		if (new_selected_idx > p_idx) {
+			set_selected_idx(new_selected_idx - 1);
+		} else if (new_selected_idx < p_idx) {
+			set_selected_idx(new_selected_idx);
+		} else {
+			set_selected_idx(-1);
+		}
+	} else {
+		// Grabbing mode is GRAB_ADD, so the point was "temporary". No undo/redo.
+		set_selected_idx(-1);
+		editing = true;
+		gradient->remove_point(p_idx);
+		_gradient_update();
+		editing = false;
+	}
+}
+
+void GradientEditorRect::set_point_offset(int p_idx, float p_offset) {
+	ERR_FAIL_COND(p_idx < 0);
+
+	// Use initial_grab_offset to determine things for the undo/redo.
+	float old_offset = initial_grab_offset;
+	if (old_offset == p_offset) {
+		return;
+	}
+
+	int old_idx = 0;
+	while (old_idx < points.size() - 1 && points[old_idx].offset < old_offset) {
+		old_idx++;
+	}
+	if (p_offset < old_offset) {
+		old_idx--; // Discount the selected point itself if old offset is bigger.
+	}
+
+	int new_idx = 0;
+	while (new_idx < points.size() - 1 && points[new_idx].offset < p_offset) {
+		new_idx++;
+	}
+
+	points.write[p_idx].offset = p_offset;
+	points.sort();
+
+	editing = true;
+	gradient->set_offset(p_idx, old_offset); // Pretend the point started from its old place.
+	Ref<EditorUndoRedoManager> &undo_redo = EditorNode::get_undo_redo();
+	undo_redo->create_action(TTR("Modify Gradient Point"), UndoRedo::MERGE_ENDS);
+	undo_redo->add_do_method(*gradient, "set_offset", old_idx, p_offset);
+	undo_redo->add_do_method(this, "_gradient_update");
+	undo_redo->add_do_method(this, "set_selected_idx", new_idx);
+	undo_redo->add_undo_method(*gradient, "set_offset", new_idx, old_offset);
+	undo_redo->add_undo_method(this, "_gradient_update");
+	undo_redo->add_undo_method(this, "set_selected_idx", old_idx);
+	undo_redo->commit_action();
+	editing = false;
+}
+
+void GradientEditorRect::set_point_color(int p_idx, const Color &p_color) {
+	ERR_FAIL_COND(p_idx < 0);
+
+	Color old_color = gradient->get_color(p_idx);
+	if (old_color == p_color) {
+		return;
+	}
+	points.write[p_idx].color = p_color;
+
+	editing = true;
+	Ref<EditorUndoRedoManager> &undo_redo = EditorNode::get_undo_redo();
+	undo_redo->create_action(TTR("Modify Gradient Point"), UndoRedo::MERGE_ENDS);
+	undo_redo->add_do_method(*gradient, "set_color", p_idx, p_color);
+	undo_redo->add_do_method(this, "_gradient_update");
+	undo_redo->add_undo_method(*gradient, "set_color", p_idx, old_color);
+	undo_redo->add_undo_method(this, "_gradient_update");
+	undo_redo->commit_action();
+	editing = false;
+}
+
+void GradientEditorRect::reverse_gradient() {
+	// I don't know why, I don't want to know why, I shouldn't have to wonder why, but for whatever reason,
+	// no matter how I tried to put this in an undo/redo structure, things broke.
+	// FIXME
```

Comment:
You could move the actual reversing to a new method like `_reverse_gradient`, bind that method (in `_bind_methods`) so it could be used by `UndoRedo`, and in here you'd just create (and commit) an UndoRedo action with just `_reverse_gradient` call in do/redo.

Seems simpler than somehow converting `Vector<Gradient::Point>` to `Variant` so you could pass it to UndoRedo.

Reply:
Doing this causes the button to become unusable for like 0.5sec, and it also only allows to go one step back. Still better than what we have right now, so I'm pushing it in a sec.

```
void GradientEditorRect::_reverse_gradient_bind() {
	editing = true;
	gradient->reverse();
	set_points(gradient->get_points());
	_gradient_update();
	editing = false;

	// Flip the selected point.
	if (selected_idx != -1) {
		set_selected_idx(points.size() - selected_idx - 1);
	}
	queue_redraw();
}

void GradientEditorRect::reverse_gradient() {
	Ref<EditorUndoRedoManager> &undo_redo = EditorNode::get_undo_redo();
	undo_redo->create_action(TTR("Reverse Gradient"), UndoRedo::MERGE_ENDS);
	undo_redo->add_do_method(this, "_reverse_gradient_bind");
	undo_redo->add_undo_method(this, "_reverse_gradient_bind");
	undo_redo->commit_action();
}
```

https://user-images.githubusercontent.com/85438892/210160139-6cb6c10a-c1e7-4106-8215-8b9e6552cfcf.mp4

Reply:
> Doing this causes the button to become unusable for like 0.5sec, and it also only allows to undo/redo to one step back.

Isn't it caused by the `UndoRedo::MERGE_ENDS` mode? :thinking: Try `UndoRedo::MERGE_DISABLED` instead.

Reply:
Oh, possible! I'll investigate. It's my first time with undo/redo systems.

Reply:
Yep. Never even thought it might be caused by this. Oh my god, it's all fixed now!